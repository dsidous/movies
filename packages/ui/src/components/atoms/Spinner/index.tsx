import React from 'react';

import useStyles from './styles';

const Spinner = () => {
  const classes = useStyles();
  return (
    <div className={classes.root}>
      <i className="fa fa-spinner fa-pulse fa-3x fa-fw" />
      <span className="sr-only">Loading...</span>
    </div>
  );
};

export default Spinner;
