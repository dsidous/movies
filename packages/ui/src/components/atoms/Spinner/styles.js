import { makeStyles } from '@material-ui/core/styles';

export default makeStyles(
  {
    root: {
      display: 'grid',
      gridTemplateRows: '100vh',
      width: '100%',

      '& i': {
        alignSelf: 'center',
        justifySelf: 'center',
      },
    },
  },
  { name: 'spinner' },
);
