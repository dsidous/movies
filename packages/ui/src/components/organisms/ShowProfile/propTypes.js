import { array, object, func } from 'prop-types';

export const propTypes = {
  config: object.isRequired,
  show: object.isRequired,
  dcolor: array.isRequired,
  handleShowClick: func.isRequired,
  handleFullCrewClick: func.isRequired,
};

export const defaultProps = {};
