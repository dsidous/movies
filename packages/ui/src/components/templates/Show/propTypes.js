import { object, bool } from 'prop-types';

export const propTypes = {
  config: object,
  show: object,
  loading: bool,
};

export const defaultProps = {};
