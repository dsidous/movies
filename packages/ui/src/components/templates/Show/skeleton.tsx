import React from 'react';
import Skeleton from '@material-ui/lab/Skeleton';
import { Container, Typography, makeStyles } from '@material-ui/core';

const useStyles = makeStyles(theme => ({
  header: {
    [theme.breakpoints.down('xs')]: {
      display: 'none',
    },
  },
  main: {
    display: 'flex',
    marginTop: '24px',
  },
  cast: {
    justifyContent: 'space-between',
    display: 'flex',
    marginTop: '24px',
    height: '96px',
    width: '100%',
  },
  castBox: {
    height: '50px',
    width: '40%',

    [theme.breakpoints.up('sm')]: {
      height: '100%',
      width: '20%',
    },
  },
}));

export default () => {
  const classes = useStyles();
  return (
    <Container>
      <Skeleton
        variant="rect"
        width="100%"
        height={200}
        className={classes.header}
      />
      <Container className={classes.main}>
        <Skeleton variant="rect" width="100%" height={200} />
        <Container>
          <Typography variant="h1">
            <Skeleton />
          </Typography>
          <Skeleton />
          <Skeleton />
          <Skeleton />
          <Skeleton width="70%" />
        </Container>
      </Container>
      <Container className={classes.cast}>
        <Skeleton variant="rect" className={classes.castBox} />
        <Skeleton variant="rect" className={classes.castBox} />
        <Skeleton variant="rect" className={classes.castBox} />
        <Skeleton variant="rect" className={classes.castBox} />
      </Container>
    </Container>
  );
};
