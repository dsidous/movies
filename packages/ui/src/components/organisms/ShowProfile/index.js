/* eslint-disable camelcase */
import React from 'react';
import Style from 'style-it';

import { propTypes } from './propTypes';
import PlayTrailer from '../../atoms/PlayTrailer';
import TopCast from '../../molecules/TopCast';
import Crew from '../../atoms/Crew';
import SimilarMovies from '../SimilarMovies';
import FullScreenBackdrop from '../../atoms/FullScreenBackdrop';
import Reviews from '../../molecules/Reviews';
import SEO from '../../atoms/SEO';
import WatchlistBookmark from '../../atoms/WatchlistBookmark';
import LastSeason from '../../molecules/LastSeason';

const ShowProfile = ({
  config: {
    images: { secure_base_url, poster_sizes, backdrop_sizes },
  },
  show: {
    id,
    backdrop_path,
    poster_path,
    genres,
    name,
    title,
    release_date,
    first_air_date,
    reviews,
    runtime,
    vote_average,
    tagline,
    overview,
    images: { backdrops },
    videos,
    similar,
    credits: { cast, crew },
    seasons,
  },
  dcolor,
  show,
  handleFullCrewClick,
  handleShowClick,
}) => {
  const showDate = first_air_date || release_date;
  const showTitle = title || name;
  let lastSeason = null;

  if (seasons) {
    [lastSeason] = seasons.sort((a, b) => b.season_number - a.season_number);
  }

  const posterURL = secure_base_url + poster_sizes[3] + poster_path;
  const backdropURL = secure_base_url + backdrop_sizes[1] + backdrop_path;
  const video = videos
    ? videos.filter(element => element.type === 'Trailer')[0]
    : [];

  const genres_html = genres.map(genre => (
    <li key={genre.id} className="movie-genres">
      {genre.name}
    </li>
  ));

  return (
    <div>
      <Style>
        {`
        .main-header:after {
          background-image: radial-gradient(at 10% 30%, rgb(${dcolor[0]},${dcolor[1]},${dcolor[2]}) 0%, #342931 100%);
        }
        .main-header:before {
          background-image: url(${backdropURL});
        }
      `}
      </Style>
      <SEO title={showTitle} />
      <div className="full-background">
        {backdrops[0] && (
          <FullScreenBackdrop
            backdrops={backdrops.map(
              image =>
                `${secure_base_url}${backdrop_sizes[2]}${image.file_path}`,
            )}
          />
        )}
      </div>

      <div className="main-header">
        <div className="main-header-inner">
          <div className="poster">
            {poster_path !== null ? (
              <img src={posterURL} alt="poster" />
            ) : (
              <div className="movie-no-image-holder" />
            )}
            <WatchlistBookmark movie={show} />
          </div>
          <div className="movie-data">
            <h1 className="movie-title">
              {showTitle}
              <span className="movie-rating">{vote_average}</span>
            </h1>

            <ul className="title-tags">
              <li>{showDate.slice(0, 4)}</li>
              {runtime && <li>{`${runtime} min`}</li>}
              <li>
                <ul className="title-tags__genres">{genres_html}</ul>
              </li>
            </ul>

            <h4 className="movie-tagline">{tagline}</h4>
          </div>

          <div className="overview">
            <h4>Overview</h4>
            {overview}
          </div>

          {video && <PlayTrailer video={video} />}

          {crew[0] && <Crew crew={crew.slice(0, 4)} />}
        </div>
      </div>
      {cast[0] && (
        <TopCast
          cast={cast.slice(0, 6)}
          handleFullCrewClick={handleFullCrewClick}
        />
      )}
      {lastSeason && <LastSeason tvId={id} season={lastSeason} />}

      {reviews && reviews[0] && <Reviews reviews={reviews} />}

      {similar.results[0] && (
        <SimilarMovies
          similar={similar.results}
          handleMovieClick={handleShowClick}
        />
      )}
    </div>
  );
};

ShowProfile.propTypes = propTypes;

export default ShowProfile;
