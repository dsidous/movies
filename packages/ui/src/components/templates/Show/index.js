/* eslint-disable camelcase */
import React, { useState, useEffect } from 'react';
import * as Vibrant from 'node-vibrant';
import { useHistory } from 'react-router-dom';

import { propTypes } from './propTypes';
import Spinner from '../../atoms/Spinner';
import PageTransition from '../../atoms/PageTransition/index';
import ShowProfile from '../../organisms/ShowProfile';

const Show = ({ loading, show, config }) => {
  const history = useHistory();
  const media = show.name ? 'tv' : 'movie';

  const [dcolor, setDcolor] = useState([0, 0, 0]);

  const handleShowClick = showId => {
    history.push(`/${media}/${showId}`);
  };

  const handleFullCrewClick = () => {
    const { id } = show;
    history.push(`/${media}/${id}/crew`);
  };

  useEffect(() => {
    const getPalette = async () => {
      const { poster_path } = show;
      const { images } = config;
      const path = images.base_url + images.poster_sizes[3] + poster_path;
      const palette = await Vibrant.from(path).getPalette();
      // eslint-disable-next-line no-underscore-dangle
      await setDcolor(palette.DarkVibrant._rgb);
    };

    if (!loading) {
      getPalette();
    }
  }, [loading, config, show]);

  if (loading) {
    return <Spinner />;
  }

  return (
    <div>
      <PageTransition>
        <ShowProfile
          key={show.id}
          config={config}
          show={show}
          dcolor={dcolor}
          handleShowClick={handleShowClick}
          handleFullCrewClick={handleFullCrewClick}
        />
      </PageTransition>
    </div>
  );
};

Show.propTypes = propTypes;

export default Show;
