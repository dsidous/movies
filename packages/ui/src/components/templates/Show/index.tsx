import React, { useState, useEffect } from 'react';
import Vibrant from 'node-vibrant';

import PageTransition from '../../atoms/PageTransition/index';
import ShowProfile from '../../organisms/ShowProfile';
import Skeleton from './skeleton';
import useConfig from '../../hooks/useConfig';

import { Show } from '@typesRoots/show';

interface Props {
  loading: boolean;
  show: Show;
}

const Show: React.FC<Props> = ({ loading, show }) => {
  const { getImageURL } = useConfig();

  const [dcolor, setDcolor] = useState([0, 0, 0]);

  useEffect(() => {
    const getPalette = async () => {
      const { poster_path } = show;
      const path = getImageURL({
        filePath: poster_path,
        mediaType: 'poster',
        size: 0,
      });
      const palette = await Vibrant.from(path).getPalette();
      setDcolor(palette.DarkVibrant!.getRgb());
    };

    if (!loading && show.poster_path !== null) {
      getPalette();
    }
  }, [loading, show]);

  return (
    <div>
      <PageTransition>
        {loading ? (
          <Skeleton />
        ) : (
          <ShowProfile key={show.id} show={show} dcolor={dcolor} />
        )}
      </PageTransition>
    </div>
  );
};

export default Show;
