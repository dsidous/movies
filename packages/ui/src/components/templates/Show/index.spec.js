import React from 'react';
import { useHistory as useHistoryMock } from 'react-router-dom';
import { shallow } from 'enzyme';

import ShowProfile from '../../organisms/ShowProfile';
import Show from '.';
import Spinner from '../../atoms/Spinner';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: jest.fn(),
}));

const mockProps = {
  config: {
    images: {
      secure_base_url: 'path/to/image',
      poster_sizes: ['1', '2', '3', '4'],
      backdrop_sizes: ['1', '2', '3', '4'],
    },
  },
  show: {
    id: 1,
    title: 'title',
  },
};

describe('Templates/Show', () => {
  const setHistory = jest.fn();

  useHistoryMock.mockImplementation(() => ({
    push: setHistory,
  }));

  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<Show {...mockProps} />);
  });

  afterEach(() => {
    setHistory.mock.calls = [];
  });

  it('should render as expected', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a spinner when loading', () => {
    wrapper = shallow(<Show {...mockProps} loading />);

    expect(wrapper.find(Spinner).length).toBe(1);
  });

  it('should handle show click', () => {
    const newMovieId = 2;

    wrapper.find(ShowProfile).props().handleShowClick(newMovieId);

    expect(setHistory.mock.calls[0][0]).toEqual(`/movie/${newMovieId}`);
  });

  it('should handle full crew click', () => {
    wrapper.find(ShowProfile).props().handleFullCrewClick();

    expect(setHistory.mock.calls[0][0]).toEqual(
      `/movie/${mockProps.show.id}/crew`,
    );
  });
});
