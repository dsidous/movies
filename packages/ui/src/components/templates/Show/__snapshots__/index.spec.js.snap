// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Templates/Show should render as expected 1`] = `ShallowWrapper {}`;
